api/sendfild/service/waitmanager/waitrunner.go

```diff
@@ -0,0 +1,188 @@
+package waitmanager
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/filecoin-project/lotus/api/apistruct"
+	"github.com/ipfs/go-cid"
+	"github.com/textileio/powergate/v2/lotus"
+	"github.com/textileio/textile/v2/api/sendfild/service/store"
+)
+
+type WaitResult struct {
+	LatestMessageCid string
+	Err              error
+}
+
+type WaitRunner struct {
+	messageCid       string
+	confidence       uint64
+	waitTimeout      time.Duration
+	store            *store.Store
+	lotusClient      *apistruct.FullNodeStruct
+	closeLotusClient func()
+	listeners        []chan WaitResult
+	lock             sync.Mutex
+	mainCtx          context.Context
+	mainCtxCancel    context.CancelFunc
+	waitCtx          context.Context
+	waitCtxCancel    context.CancelFunc
+}
+
+func NewWaitRunner(ctx context.Context, messageCid string, confidence uint64, waitTimeout time.Duration, store *store.Store, cb lotus.ClientBuilder) (*WaitRunner, error) {
+	mainCtx, mainCtxCancel := context.WithCancel(ctx)
+
+	w := &WaitRunner{
+		messageCid:    messageCid,
+		confidence:    confidence,
+		waitTimeout:   waitTimeout,
+		store:         store,
+		mainCtx:       mainCtx,
+		mainCtxCancel: mainCtxCancel,
+	}
+
+	var err error
+	w.lotusClient, w.closeLotusClient, err = cb(mainCtx)
+	if err != nil {
+		return nil, fmt.Errorf("creating lotus client: %v", err)
+	}
+
+	return w, nil
+}
+
+func (w *WaitRunner) Start() {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if w.waitCtxCancel != nil {
+		return
+	}
+	w.waitCtx, w.waitCtxCancel = context.WithTimeout(w.mainCtx, w.waitTimeout)
+	w.waitAndNotify()
+}
+
+type CancelListenerFunc = func()
+
+func (w *WaitRunner) AddListener(listener chan WaitResult) CancelListenerFunc {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.listeners = append(w.listeners, listener)
+	return func() {
+		w.lock.Lock()
+		defer w.lock.Unlock()
+		index := -1
+		for i, l := range w.listeners {
+			if l == listener {
+				index = i
+				break
+			}
+		}
+		if index >= 0 {
+			// ToDo: Write unit test to make sure adding multiple listeners and removing some works for remaining listeners.
+			w.listeners[index] = w.listeners[len(w.listeners)-1]
+			w.listeners = w.listeners[:len(w.listeners)-1]
+		}
+	}
+}
+
+func (w *WaitRunner) Close() error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.closeLotusClient()
+	w.waitCtxCancel()
+	w.mainCtxCancel()
+	return nil
+}
+
+func (w *WaitRunner) waitAndNotify() {
+	go func() {
+		c, err := cid.Decode(w.messageCid)
+		if err != nil {
+			if err := w.store.FailTxn(w.mainCtx, w.messageCid, err.Error()); err != nil {
+				w.notifyErr(fmt.Errorf("failing txn in store: %v", err))
+				return
+			}
+			w.notify(w.messageCid)
+			return
+		}
+
+		if err := w.store.SetWaiting(w.mainCtx, w.messageCid, true); err != nil {
+			w.notifyErr(fmt.Errorf("setting txn to waiting: %v", err))
+			return
+		}
+
+		res, err := w.lotusClient.StateWaitMsg(w.waitCtx, c, w.confidence)
```

Comment:
Notice the use of `waitCtx` here where elsewhere we use `mainCtx`. This is because the `waitCtx` is set to a reasonable timeout, and we want that, and in the case it does time out, we need the `mainCtx` to do things like update the `Store`.